dhall-lsp-server/src/Dhall/LSP/Backend/Freezing.hs

```diff
@@ -0,0 +1,56 @@
+module Dhall.LSP.Backend.Freezing (
+  computeSemanticHash,
+  getAllImportsWithHashPositions,
+  getImportHashPosition,
+  stripHash
+) where
+
+import Dhall.Parser (Src(..))
+import Dhall.Core (Expr(..), Import(..), ImportHashed(..), subExpressions)
+
+import Control.Lens (universeOf)
+import Data.Text (Text)
+import qualified Data.Text as Text
+
+import Dhall.LSP.Backend.Dhall (FileIdentifier, Cache, DhallError, typecheck,
+  normalize, hashNormalToCode, load)
+import Dhall.LSP.Backend.Diagnostics (Range(..), rangeFromDhall, positionFromMegaparsec, positionToOffset, relativePosition)
+import Dhall.LSP.Backend.Parsing (getImportHash)
+
+-- | Given an expression (potentially still containing imports) compute its
+-- 'semantic' hash in the textual representation used to freeze Dhall imports.
+computeSemanticHash :: FileIdentifier -> Expr Src Import -> Cache ->
+  IO (Either DhallError (Cache, Text))
+computeSemanticHash fileid expr cache = do
+  loaded <- load fileid expr cache
+  case loaded of
+    Left err -> return (Left err)
+    Right (cache', expr') -> case typecheck expr' of
+      Left err -> return (Left err)
+      Right (wt,_) -> do
+        return (Right (cache', hashNormalToCode (normalize wt)))
+
+stripHash :: Import -> Import
+stripHash (Import (ImportHashed _ importType) mode) =
```

Comment:
This is outside of the scope of this pull request, but at some point we should add some lenses for dealing with the deeply nested `Import*` hierarchy, so that this could be factored into:

```haskell
strip lens = set lens empty

hash = importHashedLens . hashLens

stripHash = strip hash
```

Reply:
This would be amazing, we have some pretty verbose pattern matches in Spago that we could totally get rid of if we had lenses (e.g. see [here](https://github.com/spacchetti/spago/blob/3eaba105a0a9ff59a35709e5d6336c1d81d73215/src/Spago/PackageSet.hs#L144))